xla/service/scatter_determinism_expander.cc

```diff
@@ -0,0 +1,459 @@
+
+  const Shape& scalar_index_shape =
+      ShapeUtil::MakeShape(indices_shape.element_type(), {num_indices});
+
+  auto* comparison = ScalarSortingComparison(
+      scatter->GetModule(),
+      ShapeUtil::MakeShape(indices_shape.element_type(), {}),
+      ShapeUtil::MakeShape(updates_shape.element_type(), {}),
+      scatter_updates.size());
+
+  std::vector<HloInstruction*> sort_operands = {scalar_indices};
+  std::vector<Shape> sort_shapes = {scalar_index_shape};
+  for (auto update : scatter_updates) {
+    sort_operands.push_back(update);
+    sort_shapes.push_back(update->shape());
+  }
+
+  auto* sorting = parent->AddInstruction(HloInstruction::CreateSort(
+      ShapeUtil::MakeTupleShape(sort_shapes), 0, sort_operands, comparison,
+      false /*is_stable*/));
+  auto* sorted_scalar_indices =
+      parent->AddInstruction(HloInstruction::CreateGetTupleElement(
+          scalar_indices->shape(), sorting, 0));
+
+  std::vector<HloInstruction*> sorted_updates(scatter_updates.size());
+  for (int i = 0; i < scatter_updates.size(); i++) {
+    sorted_updates[i] =
+        parent->AddInstruction(HloInstruction::CreateGetTupleElement(
+            scatter_updates[i]->shape(), sorting, i + 1));
+  }
+  std::vector<HloInstruction*> sorted_tensors = {sorted_scalar_indices};
+  sorted_tensors.insert(sorted_tensors.end(), sorted_updates.begin(),
+                        sorted_updates.end());
+  return sorted_tensors;
+}
+
+// CreateScanWithIndices performs a prefix scan operation (akin to parallel
+// prefix sum) on the updates and indices, to compute the accumulated updates in
+// log(n) time.
+//
+// High-level algorithm:
+//
+// Iteration through log2(num_updates):
+//   - For each iteration, the `updates` tensor will be sliced and padded to
+//   perform shifting by `offset`.
+//   - Similarly, the `indices` tensor is also sliced and padded.
+//   - A mask is created that compares each element of shifted `indices` and
+//   original `indices` are equal (used to avoid combining updates from
+//   different indices).
+//   - The `to_apply` function is used to combine the original and shifted
+//   updates to generate a combined update tensor.
+//   - Based on the mask, the new update tensor will choose from either the
+//   combined update or the original update.
+//   - The result becomes the `new_updates`, which is then used as the
+//   input for the next iteration.
+static StatusOr<HloInstruction*> CreateScanWithIndices(
+    HloComputation* parent, HloInstruction* updates, HloInstruction* indices,
+    HloComputation* to_apply) {
+  const Shape& updates_shape = updates->shape();
+  const Shape& indices_shape = indices->shape();
+  // Get the length of the input array
+  int64_t num_updates = updates_shape.dimensions(0);
+
+  // Calculate the number of iterations needed (log_2(n))
+  int64_t log_n = static_cast<int64_t>(std::ceil(std::log2(num_updates)));
+
+  // Placeholder for offset calculation (2^d)
+  int64_t offset;
```

Comment:
I don't see an advantage to declare the variable here, instead of in the loop where it is assigned. The compiler should be able to do this optimization, and it seems easier to read to move the declaration down.

Reply:
changed